neuralpp/symbolic/normalizer.py

```diff
@@ -1,108 +1,10 @@
-    return None
-
-
-def first_non_constant_formula_of_expression(expression: Expression) -> Optional[Expression]:
-    """ returns the first non_constant_formula occurring in a given expression in a breadth-first search. """
-    return _bfs_first_non_constant_formula_of_expression(expression, deque())
-
-
-class Normalizer:
-    simplifier = ContextSimplifier()
 
-    @staticmethod
-    def _basic_normalize(expression: Expression, context: Z3SolverExpression) -> Expression:
-        """
-        The word `basic` in the function name is not to be confused with that in `BasicExpression`.
-        Here it refers to normalization of non-quantifier expressions.
-        The function assumes context is satisfiable, otherwise will raise (from Normalizer.simplifier.simplify()).
-        """
-        simplified_expression = Normalizer.simplifier.simplify(expression, context)
-        # We look for non-constant formula, because we should not split on constants (i.e., True/False)
-        # as it is meaningless and would cause the algorithm to not terminate.
-        # E.g., split on `True` would result in `if True then True else False`.
-        first_non_constant_formula = first_non_constant_formula_of_expression(simplified_expression)
-        if first_non_constant_formula is None:
-            return simplified_expression
-
-        context1 = context & first_non_constant_formula
-        # The following statement can be proved by contradiction:
-        # if not isinstance(context1, Z3SolverExpression), then context & first_non_constant_formula must be False,
-        # then context -> ~first_non_constant_formula, then in Normalizer.simplifier.simplify(expression, context),
-        # first_non_constant_formula as a subexpression of expression should have been replaced by False,
-        # since its negation is implied by context, thus should not be a subexpression of simplified_expression. Q.E.D.
-        assert isinstance(context1, Z3SolverExpression)
-        expression1 = simplified_expression.replace(first_non_constant_formula, basic_true)
-        then_clause = Normalizer._normalize(expression1, context1)
-
-        context2 = context & ~first_non_constant_formula
-        assert isinstance(context2, Z3SolverExpression)
-        expression2 = simplified_expression.replace(first_non_constant_formula, basic_false)
-        else_clause = Normalizer._normalize(expression2, context2)
-
-        return if_then_else(first_non_constant_formula, then_clause, else_clause)
-
-    @staticmethod
-    def _normalize(expression: Expression, context: Z3SolverExpression) -> Expression:
-        """ The function assumes context is satisfiable, otherwise will raise. """
-        if isinstance(expression, QuantifierExpression):
-            return Normalizer._quantifier_expression_normalize(expression, context)
-        else:
-            return Normalizer._basic_normalize(expression, context)
 
+class Normalizer(ABC):
```

Comment:
changed `Normalizer` to be an interface class. The old implementation is moved to `QuantifierFreeNormalizer`